TASK: Check Neos.Demo refactoring for ES CR specifics

ahaeslich opened the pull request:
- New dimension configuration format incl. dimension value mapping to url path segment
- Use Neos.Caching.entryIdentifierForNode() helper in Fusion
- Check usage of `Neos.Node.inBackend(node)` instead of `node.context.inBackend`
- Use `Neos.Dimension` helper to render language menu in refactored header

ahaeslich committed:
TASK: Reapply new dimension configuration format incl. dimension value mapping to url path segment

ahaeslich committed:
TASK: Reapply use Neos.Caching.entryIdentifierForNode() helper in Fusion

ahaeslich committed:
TASK: Reapply usage of `Neos.Node.inBackend(node)` instead of `node.context.inBackend`

ahaeslich committed:
TASK: Use `Neos.Dimension` helper to render language menu in refactored header

To activate the language in the dropdown menu the refactored header relies on the generation of alternate language links which are generated via `Neos.Seo` package.

Relates: neos/neos-seo#171

mhsdesign commented:
help

ahaeslich commented on Resources/Private/Fusion/Integration/Header.fusion:
```diff
@@ -18,7 +18,13 @@ prototype(Neos.Demo:Integration.Header.Renderer) < prototype(Neos.Demo:Presentat
-    currentLanguage = ${site.context.dimensions.language[0]}
+    currentLanguage = ${Neos.Dimension.currentValue(site, 'language')}
+    languages = Neos.Fusion:Map {
+        items = ${Neos.Dimension.all(site)['language'].values.iterator}
```
This works but there are two issues I have with this line of code:

- In neos/neos-development-collection#4085 we removed the `findPotentialDimensionValues` function but this is now the second time of the usage `${Neos.Dimension.all(site)['language']}` so I guess we should reintroduce it. Maybe with a better naming ;)
- If we have a collection object at hand (`values` is of type `ContentDimensionValues`) do we want to see this usage of `iterator` to work with a e.g. `Neos.Fusion:Map` or similar Implementations which as of now only accept an `array`.

Any thoughts @skurfuerst / @bwaidelich / @mficzel ?

mficzel replied:
1. I also think we will introduce a permutation of "findPotentialDimensionValues" (all values) again eventually. Probably in distinction to "findDimensionRootValue" (values without fallback). Better names would really help.
2. We should make `Map` accept an `iterator` that does not implement `countable`. The only value we cannot calculate in this case is `iterator.isLast`.

bwaidelich replied:
I didn't look into it very closely, but IMO this Fusion code is _far_ too low level.
Couldn't we add a dedicated helper, like
```
languages = ${Neos.Dimension.nameToBeFound(site, 'language')}
```
?

mficzel replied:
Looping/Mapping over uncountable collections seems to be rather simple see https://github.com/neos/neos-development-collection/pull/4168 ... however making collections countable would often make sense.

skurfuerst replied:
(just discussing with @ahaeslich )

suggestion / idea:

```
// "language" would be ContentDimensionId
languages = ${Neos.Dimension.allDimensionValues(site, 'language')}
```

-> should return `ContentDimensionValues` object I'd say -> this is iterable.

bwaidelich replied:
@skurfuerst yes, I like that a lot!

ahaeslich replied:
Will add that so we can proceed with this MR and Neos.Seo

ahaeslich commented:
> help

Better @mhsdesign 😉?

ahaeslich committed:
TASK: Remove icon prefix

Thanks @dlubitz for noticing

ahaeslich committed:
TASK: Use `allDimensionValues` to render language menu

Relates: https://github.com/neos/neos-development-collection/pull/4244

ahaeslich committed:
TASK: Update Neos.Demo content to work with `./flow cr:import`

- Remove old exported content format used by `./flow site:import`

ahaeslich committed:
BUGFIX: Fix label of current language in language menu

After refactoring of `${Neos.Dimension.currentValue()}` `currentLanguage` now returns a `ContentDimensionValue` instead of a string. Therefore, we can use `{props.currentLanguage.configuration.label}` in fusion to output the correct label.

Needs: https://github.com/neos/neos-development-collection/pull/4249

kdambekalns reviewed (approved):
✅ by 👀
